design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,140 @@
+More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)
+It is projected that COSI will be the only supported object storage driver in the near future.
+
+## COSI Driver Deployment
+
+The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+### End to End Workflow
+
+#### Changes in Common.yaml
+
+Following contents need to be append to `common.yaml` :
+
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>
+
+#### cephcosi.ceph.rook.io CRD
+
+The users need to define following CRD to bring up the COSI driver for the ceph object store.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCOSI
+metadata:
+  name: rook-ceph-cosi
+  namespace: rook-ceph
+spec:
+  objectStoreName: <object-store-name>
+```
+
+The current COSI architecture allows driver to interact with one object store endpoint with limitations in the APIs. Hence planning to support one object store for the driver in Rook 1.12. When multiple object stores are supported in the future, the user can define `spec.objectStoreName` to `all`. Rook Operator allows one such CRD in the cluster aka one ceph cosi driver for the cluster. In future this driver can support multiple object stores.
```

Comment:
Why do we need `spec.objectStoreName: all`? Is that a setting in the ceph COSI driver? Can we make that the default instead of requiring the user to specify it?